gnomad/utils/constraint.py

```diff
@@ -360,3 +412,285 @@ def collapse_strand(
+        - observed_variants - the number of observed variants in the dataset for each
+        variant. Note that the term "variant" here refers to a specific substitution,
+        context, methylation level, and coverage combination
+        - downsampling_counts_{pop} (optional) - array of observed variant counts per
+        population after downsampling. Used only when `pops` is specified.
+        - mu_snp - mutation rate
+        - possible_variants - the number of possible variants in the dataset for each
+        variant
+
+    :param coverage_ht: Input coverage Table.
+    :param weighted: Whether to weight the plateau models (a linear regression
+      model) by 'possible_variants'. Default is False.
+    :param pops: List of populations used to build plateau models.
+      Default is ().
+    :param keys: Annotations used to group observed and possible variant counts.
+      Default is ("context","ref", "alt", "methylation_level", "mu_snp").
+    :param cov_cutoff: Median coverage cutoff. Sites with coverage above this cutoff
+      are considered well covered and will be used to build plateau models. Sites below
+      this cutoff have low coverage and will be used to build coverage models. Defaults
+      to `COVERAGE_CUTOFF`.
+    :return: Coverage model and plateau models.
+    """
+    # Filter to sites with coverage above `cov_cutoff`.
+    high_cov_ht = coverage_ht.filter(coverage_ht.exome_coverage >= cov_cutoff)
+    agg_expr = {
+        "observed_variants": hl.agg.sum(high_cov_ht.observed_variants),
+        "possible_variants": hl.agg.sum(high_cov_ht.possible_variants),
+    }
+    for pop in pops:
+        agg_expr[f"observed_{pop}"] = hl.agg.array_sum(
+            high_cov_ht[f"downsampling_counts_{pop}"]
+        )
+
+    # Generate a Table with all necessary annotations (x and y listed above)
+    # for the plateau models.
+    high_cov_group_ht = high_cov_ht.group_by(*keys).aggregate(**agg_expr)
+    high_cov_group_ht = annotate_mutation_type(high_cov_group_ht)
+
+    # Build plateau models.
+    plateau_models_agg_expr = build_plateau_models(
+        cpg_expr=high_cov_group_ht.cpg,
+        mu_snp_expr=high_cov_group_ht.mu_snp,
+        observed_variants_expr=high_cov_group_ht.observed_variants,
+        possible_variants_expr=high_cov_group_ht.possible_variants,
+        pop_observed_variants_exprs_arr=hl.array(
+            [high_cov_group_ht[f"observed_{pop}"] for pop in pops]
+        ),
+        weighted=weighted,
+    )
+    plateau_models = dict(
+        high_cov_group_ht.aggregate(hl.struct(**plateau_models_agg_expr))
+    )
+
+    if pops:
+        # Map the models to their corresponding populations.
+        pop_models_arr = plateau_models["pop"]
+        for idx, pop in enumerate(pops):
+            plateau_models[pop] = pop_models_arr[idx]
+        plateau_models.pop("pop")
+    plateau_models = hl.struct(**plateau_models)
```

Comment:
not that much less redundant, but what about:
```
_plateau_models = dict(
    high_cov_group_ht.aggregate(hl.struct(**plateau_models_agg_expr))
)

plateau_models = {
    pop: hl.literal(pop_models_expr[idx]) for idx, pop in enumerate(pops),
}
plateau_models["total"] =  _plateau_models["total"],
plateau_models = hl.struct(**plateau_models)
```
I've also renamed `pop_models_arr` to `pop_models_expr` to make it clearer